README.md
<div align="center">
<hr/>
<h1>OrpailleCC</h1>
<a href="https://travis-ci.org/big-data-lab-team/OrpailleCC"><img src="https://travis-ci.org/big-data-lab-team/OrpailleCC.svg?branch=master"></a>
<a href="https://www.codacy.com/app/glatard/OrpailleCC?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=big-data-lab-team/OrpailleCC&amp;utm_campaign=Badge_Grade"><img src="https://api.codacy.com/project/badge/Grade/2746a0cd00d34441a6f7dcf1558f1089"/></a>
<a href="https://zenodo.org/badge/latestdoi/170758788"><img src="https://zenodo.org/badge/170758788.svg" alt="DOI"></a>
</div>	

OrpailleCC is data stream library written in C++. It provides a consistent
collection of data stream algorithms for embedded devices. The goal of
OrpailleCC is to support research on data stream mining for connected objects,
by facilitating the comparison and benchmarking of algorithms in a consistent
framework. It also enables programmers of embedded systems to use
out-of-the-box algorithms with an efficient implementation.
Algorithms from OrpailleCC are based on C++ templates and does not use the STL library.

## Get started
### Hello World
Let us run a basic example with a reservoir sampling \[4] of size 3.
Save the following code in *testy.cpp*.
```cpp
#include <iostream> //Included for cout
#include "reservoir_sampling.hpp"

double randy(void){ //We need this function to provide a random number generator to ReservoirSampling.
	return (double)rand() / (double)RAND_MAX; //On systems without rand, the programmer will have to define a pseudo-random function.
}

int main(){
	char hello[] = "Hello-world!"; //Create a stream
	ReservoirSampling<char, 3, randy> rs; //Instantiate a ReservoirSampling instance
	//This instance works with char, contains a reservoir of size 3 and  use the randy function to generate random numbers.
	for(int j = 0; j < 12; ++j) //Feed the ReservoirSampling instance with every element of the stream (here letters of the string)
		rs.add(hello[j]);
	for(int j = 0; j < 3; ++j) //Print every element in the reservoir
		std::cout << rs[j];
	std::cout << std::endl;
	return 0;
}
```

Then compile it with your favorite C++ compiler and run it.
```bash
$ g++ -I./src -std=c++11 testy.cpp -o testy
$ ./testy
Hll
```
### Install
#### Requirement
As the collection is designed to run on embedded system without operating
systems, OrpailleCC has very little dependencies and requirement.

- Git : to download the repository.
- C++ compiler with C++11: to compile OrpailleCC files.
- googletest: to run unit tests.
- Linux Operating System: because all instructions are given for Linux systems. However, OrpailleCC should compile properly on a Windows system as long as a C++ compiler is available.

#### Installation
To install OrpailleCC, first clone the repository. 
```bash
git clone https://github.com/big-data-lab-team/OrpailleCC.git
```
In this example, we assume that OrpailleCC is located in
`/usr/include/OrpailleCC`. Change it accordingly to your system.
```bash
ORPAILLECC_DIR=/usr/include/OrpailleCC
```

To use OrpailleCC in your project add `ORPAILLECC_DIR/src` in the include directories of the project.
Let's assume the project is the hello world example, located in *~/hello/hello.cpp*.

```cpp
#include <iostream> //Included for cout
#include <reservoir_sampling.hpp>

double randy(void){ //We need this function to provide a random number generator to ReservoirSampling.
	return (double)rand() / (double)RAND_MAX; //On systems without rand, the programmer will have to define a pseudo-random function.
}

int main(){
	char hello[] = "Hello-world!"; //Create a stream
	ReservoirSampling<char, 3, randy> rs; //Instantiate a ReservoirSampling instance
	//This instance works with char, contains a reservoir of size 3 and  use the randy function to generate random numbers.
	for(int j = 0; j < 12; ++j) //Feed the ReservoirSampling instance with every element of the stream (here letters of the string)
		rs.add(hello[j]);
	for(int j = 0; j < 3; ++j) //Print every element in the reservoir
		std::cout << rs[j];
	std::cout << std::endl;
	return 0;
}
```

To compile this code (that use the ReservoirSampling object), you need to run the following commands.

```bash
cd ~/hello
g++ -std=c++11 -I$ORPAILLECC_DIR hello.c
```

### Test
#### Unit Test
The unit tests require the `googletest` library ([Here](https://github.com/google/googletest)).
To run the unit tests, run the command `make run_test`.

#### Performance
To run a performance test on your device, compile the performance tests with
`make perf` then run `./main-perf`.

![Alt](/figures/performance.png "An example of the performance output")

#### Coverage
To observe the coverage of test function, run the following commands:
```bash
make clean
make config=debug coverage
```
These commands will clean previous object files to rebuild them with the debug options, then run the test and gather the data for the coverage.
To visualize the test coverage, simply open *coverage/index.html* into your favorite browser.

## Examples
This section provides the list of all algorithms implemented in OrpailleCC with a brief example.
### Lightweight Temporal Compression (LTC)
LTC \[0] is a compression algorithm that approximates a series of values with a linear
function. The epsilon parameter controls the amount of compression. If the
linear approximation isn't accurate enough, then a new point is
issued.

To use the LTC object, you need to include the header `ltc.hpp`.
```cpp
#include "ltc.hpp"

int main(){
	LTC<int, int, 3> comp; //Instantiate an LTC object that works with integers as element, and integers as timestamp.
	//The epsilon is set to 3. 
	for(int i = 0; i < 10000; ++i){ //10000 points are added sequentially.
		//The add function returns true, if the new point cannot be compressed with the previous ones and thus needs to be transmitted.
		bool need_transmission = comp.add(i, i%200);
		if(need_transmission){
			int timestamp, value;
			//the function get_value_to_transmit fetch the next data point to send and stores it into timestamp and value.
			comp.get_value_to_transmit(timestamp, value);
			// ... Execute the transmission or store the point.
		}
	}
}
```
### Micro-Cluster Nearest Neighbour (MC-NN)
MC-NN \[3] is a classifier based on k-nearest neighbours. It aggregates the data
points into micro-clusters and make them evolve to catch concept drifts.

```cpp
#include <iostream> //Included for cout
#include "mc_nn.hpp"

#define MCNN_FEATURE_COUNT 2
int main(){
	int const dataset_size = 9;
	/*Instantiate a MCNN classifier that uses:
		- double as features
		- which consider two features (MCNN_FEATURE_COUNT)
		- That uses 10 clusters at most
	*/
	MCNN<double, MCNN_FEATURE_COUNT, 10> classifier;
	//Initialize a fake dataset
	double dataset[][MCNN_FEATURE_COUNT] = 
						   {{ 0,  0},
							{ 1,  1},
							{ 8,  1},
							{ 1,  8},
							{ 7,  2},
							{ 2,  8},
							{-1,  0},
							{ 1,  0},
							{ 7, -1}};
	int labels[] = {1, 1, 2, 2,2,2,1,1,2};

	//Train the classifier
	for(int i = 0; i < dataset_size; ++i){
		classifier.train(dataset[i], labels[i]);
	}
	double test1[MCNN_FEATURE_COUNT] = {8, 0};
	double test2[MCNN_FEATURE_COUNT] = {-1, -1};
	int prediction1 = classifier.predict(test1);
	int prediction2 = classifier.predict(test2);
	std::cout << "Prediction { 8,  0} :" << prediction1 << std::endl;
	std::cout << "Prediction {-1, -1} :" << prediction2 << std::endl;
	return 0;
}
```
### Reservoir Sampling
The next example is the one used as a hello world example. A Reservoir 
Sample \[4] is a fixed-sized sample of the stream where all elements have 
equal probability to appear.
```cpp
#include <iostream> //Included for cout
#include "reservoir_sampling.hpp"

double randy(void){ //We need this function to provide a random number to ReservoirSampling.
	return (double)rand() / (double)RAND_MAX; //On systems without rand, the programmer will have to define a pseudo-random function.
}

int main(){
	char hello[] = "Hello-world!"; //Create a stream
	ReservoirSampling<char, 3, randy> rs; //Instantiate a ReservoirSampling instance
	//This instance works with char, contains a reservoir of size 3 and  use the randy function to generate random numbers.
	for(int j = 0; j < 12; ++j) //Feed the ReservoirSampling instance with every element of the stream (here letters of the string)
		rs.add(hello[j]);
	for(int j = 0; j < 3; ++j) //Print every element in the reservoir
		std::cout << rs[j];
	std::cout << std::endl;
	return 0;
}
```
### Chained Reservoir Sampling
The chained reservoir sampling \[1] is a variant of the reservoir sampling that allows discarding outdated data while maintaining the reservoir distribution.

```cpp
#include <iostream> //Included for cout
#include <cstdlib> //Included for malloc (note: that on other system, the malloc function may be redifined otherwise.)
#include "chained_reservoir.hpp"

//Define a structure that contains the malloc function
struct funct{
	static void* malloc(unsigned int const size){
		return std::malloc(size);
	}
};
//Declare the random function
double randy(void){
	return (double)rand() / (double)RAND_MAX;
}
#define RESERVOIR_SIZE 10
int main(){
	/*Create a ChainedReservoirSampling object:
		- That stores integer
		- Which has a reservoir size of 10 (RESERVOIR_SIZE)
		- Which uses randy as random function
		- Which uses the structure funct to call for other functions. (in that case, only memory allocation function)
	*/
	ChainedReservoirSampling<int, RESERVOIR_SIZE, randy, funct> rs;
	for(int i = 0; i < 100; ++i)
		rs.add(i, i+1); // Add new element to the reservoir (given a certain probability)
	std::cout << "Original reservoir: "; 
	for(int i = 0; i < RESERVOIR_SIZE; ++i)
		std::cout << rs[i] << " "; //Access the reservoir content
	std::cout << std::endl;
	rs.obsolete(20); //Declare all timestamp before 20 obsolete (20 is included)
	std::cout << "Obsolete reservoir: "; 
	for(int i = 0; i < RESERVOIR_SIZE; ++i)
		std::cout << rs[i] << " ";
	std::cout << std::endl;
}
```
### Bloom Filter
The Bloom filter \[5] excludes elements from the stream when they don't belong to 
a pre-defined set.
```cpp
#include <iostream> //Included for cout
#include "bloom_filter.hpp"

#define BLOOM_FILTER_SIZE 50
unsigned int hash_int(int* element){
	return (*element)%BLOOM_FILTER_SIZE;
}
int main(){
	auto p = hash_int;
	/*Instantiate a BloomFilter object:
		- The element type taken as input (int)
		- The size (in bit) of the filter
		- The number of hash function to use
	With p containing the hash function.
	*/
	BloomFilter<int, BLOOM_FILTER_SIZE, 1> bf(&p);
	//Add 3 elements to the filter
	bf.add(10);
	bf.add(7);
	bf.add(73);
	std::cout << "Element accepted by the bloom filter: ";
	for(int i = 0; i < 100; ++i){
		if(bf.lookup(i)){
			std::cout << i << " ";
		}
	}
	std::cout << std::endl;
}
```
Note that, due to the Bloom Filter size, more than three elements will be recognized by the filter.

### Cuckoo Filter
The Cuckoo filter \[2] is used when elements have to be removed from the pre-defined 
set of accepted elements.
```cpp
#include <iostream> //Included for cout
#include "cuckoo_filter.hpp"

#define CUCKOO_BUCKET_COUNT 32
#define CUCKOO_ENTRY_BY_BUCKET 6
#define CUCKOO_ENTRY_SIZE 3

//This structure contains functions for the cuckoo filter.
struct funct_cuckoo{
	//The fingerprint function should return value within the size of an entry size.
	static unsigned char fingerprint(int const* e){
		//mod 7 == value between 0 and 6, +1 == value between 1 and 7, so the empty value (0x0) is avoided
		return ((*e)%7)+1; 
	}
	//The hash function that takes an element as input and return an index.
	static unsigned int hash(int const* e){
		return (*e)%CUCKOO_BUCKET_COUNT;		
	}
	//This hash function take a fingerprint as input and return an index.
	static unsigned int hash(unsigned char fingerprint){
		//Here, we make sure that the two hashfunctions does not return the same value
		return (fingerprint * 7)%CUCKOO_BUCKET_COUNT;
	}
};
double randy(void){ //We need this function to provide a random number to CuckooFilter.
	return (double)rand() / (double)RAND_MAX; //On systems without rand, the programmer will have to define his own function.
}
int main() { 
	CuckooFilter<int, CUCKOO_BUCKET_COUNT, CUCKOO_ENTRY_BY_BUCKET, CUCKOO_ENTRY_SIZE, funct_cuckoo, randy> cf;
	cf.add(10);
	cf.add(7);
	cf.add(73);
	std::cout << "Element accepted by the cuckoo filter: ";
	for(int i = 0; i < 100; ++i){
		if(cf.lookup(i)){
			std::cout << i << " ";
		}
	}
	std::cout << std::endl;
}
```

## How can I help?
- Report issues and seek support in the Issues tab.
- Write new examples or improve existing examples and share them with a pull request. 
- Submit ideas for future algorithms to integrate.
- Submit pull requests with algorithm implementation.
- Submit pull requests with additional test cases.

## References
-  \[0] Schoellhammer, Tom and Greenstein, Ben and Osterweil, Eric and Wimbrow, Michael and Estrin, Deborah (2004), "Lightweight temporal compression of microclimate datasets"
-  \[1] Babcock, Brian and Datar, Mayur and Motwani, Rajeev (2002), "Sampling from a moving window over streaming data", Proceedings of the thirteenth annual Association for Computing Machinery-SIAM symposium on Discrete algorithms, pages 633--634
-  \[2] Fan, Bin and Andersen, Dave G and Kaminsky, Michael and Mitzenmacher, Michael (2014), "Cuckoo filter: Practically better than bloom", Proceedings of the 10th Association for Computing Machinery International on Conference on emerging Networking Experiments and Technologies, pages 75--88
-  \[3] Tennant, Mark and Stahl, Frederic and Rana, Omer and Gomes, Joao Bartolo (2017), "Scalable real-time classification of data streams with concept drift", Future Generation Computer Systems, pages 187--199
-  \[4] Vitter, Jeffrey S (1985), "Random sampling with a reservoir", Association for Computing Machinery Transactions on Mathematical Software (TOMS), pages 37--57
-  \[5] Burton H. Bloom (1970), "Space/Time Trade-offs in Hash Coding with Allowable Errors", Communications of the Association for Computing Machinery


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [Unreleased]

## [1.1] - 2020-10-27
### Added
- Hoeffding Tree
- Exponential Reservoir Sampling
- Mondrian Forest
- Naïve Bayes
- Multilayer perceptron

### Changed
- Reservoir Sampling is split in two classes
- Functions to add elements in Reservoir Sampling now return the index where the element is added.

### Fixed
- Reservoir Sampling statistical bugs

## [1.0] - 2019-06-20
### Added
- Reservoir Sampling
- Chained Reservoir Sampling
- Micro-Cluster Nearest Neighbour (MC-NN)
- Lightweight Temporal Compression (LTC)
- Bloom Filter
- Cuckoo Filter



main-performance.cpp
#include <unistd.h>
#include <iostream>
#include "bloom_filter.hpp"
#include "cuckoo_filter.hpp"
#include "ltc.hpp"
#include "reservoir_sampling.hpp"
#include "mc_nn.hpp"
#include <sys/time.h>
using namespace std;

#define BLOOM_FILTER_SIZE 600
unsigned int hash_int(int* element){
	return (*element)%BLOOM_FILTER_SIZE;
}
double When() {
    struct timeval tp;
    gettimeofday(&tp, NULL);
    return ((double) tp.tv_sec + (double) tp.tv_usec * 1e-6);
}

#define CUCKOO_BUCKET_COUNT 32
#define CUCKOO_ENTRY_BY_BUCKET 6
#define CUCKOO_ENTRY_SIZE 3
struct funct_cuckoo{
	static unsigned char fingerprint(int const* e){
		//mod 7 == value between 0 and 6, +1 == value between 1 and 7, so the empty value (0x0) is avoided
		return ((*e)%7)+1; 
	}
	static unsigned int hash(int const* e){
		return (*e)%CUCKOO_BUCKET_COUNT;		
	}
	/*
	 * Make the combination of hash element and hash fingerprint does not lead to the same value
	 * Hash of fingerprint should probably not return a 0 or a 2^CUCKOO_ENTRY_SIZE-1 because it create same h1 and h2
	 */
	static unsigned int hash(unsigned char fingerprint){
		return (fingerprint * 7)%CUCKOO_BUCKET_COUNT;
	}
};
double randy(void){
	return (double)rand() / (double)RAND_MAX;
}
class functions{
	public:
	static void* malloc(int const size){
		return std::malloc(size);
	}
	static void free(void* p){
		std::free(p);
	}
	static double log(double const x){
		return std::log(x);
	}
	static double log2(double const x){
		return std::log2(x);
	}
	static double exp(double const x){
		return std::exp(x);
	}
	static double sqrt(double const x){
		return std::sqrt(x);
	}
	static bool isnan(double const x){
		return std::isnan(x);
	}
	//TODO: we need a rand_uniform and a log functions (log == ln)
	static double rand_uniform(void){
		return static_cast<double>(rand())/static_cast<double>(RAND_MAX);
	}
	static double random(void){
		return (static_cast<double>(std::rand()) / static_cast<double>(RAND_MAX));
	}
	template<class T>
	static int round(T const x){
		return std::round(x);
	}
	template<class T>
	static int floor(T const x){
		return std::floor(x);
	}
};

void test_cuckoo(void) { 
	cout << "\t=== Cuckoo ===" << endl;
	CuckooFilter<int, CUCKOO_BUCKET_COUNT, CUCKOO_ENTRY_BY_BUCKET, CUCKOO_ENTRY_SIZE, funct_cuckoo, randy> cf;
	unsigned int const count_add = 100000000;
	unsigned int const count_lookup = 100000000;
	double start = When();
	for(int i = 0; i < count_add; ++i){
		cf.add(i);
		if(i%100 == 0)
			cf.clear();
	}
	double stop = When();
	cout << "Time (Insert): " << (stop - start) << " (" << (((stop - start) / count_add) * 1e9) << " ns/item)" << endl;

	cf.clear();
	start = When();
	for(int i = 0; i < count_lookup; ++i){
		cf.lookup(i);
	}
	stop = When();
	cout << "Time (Lookup): " << (stop - start) << " (" << (((stop - start) / count_lookup) * 1e9) << " ns/item)" << endl;
}
void test_bloom(void){
	cout << "\t=== Bloom ===" << endl;
	auto p = hash_int;
	unsigned int const count_add = 1000000000;
	unsigned int const count_lookup = 1000000000;
	BloomFilter<int, BLOOM_FILTER_SIZE, 1> bf(&p);
	double start = When();
	for(int i = 0; i < count_add; ++i){
		bf.add(i);
	}
	double stop = When();
	cout << "Time (Insert): " << (stop - start) << " (" << (((stop - start) / count_add) * 1e9) << " ns/item)" << endl;

	bf.clear();
	start = When();
	for(int i = 0; i < count_lookup; ++i){
		bf.lookup(i);
	}
	stop = When();
	cout << "Time (Lookup): " << (stop - start) << " (" << (((stop - start) / count_lookup) * 1e9) << " ns/item)" << endl;
}
#define LTC_SIZE 1000000
void test_ltc(void){
	cout << "\t=== LTC ===" << endl;
	double start, stop;
	int count_smooth = 0, count_rough = 0, count_linear = 0, count_rand = 0;
	int vals_smooth[LTC_SIZE];
	int vals_rough[LTC_SIZE];
	int* vals_rand = vals_smooth;
	for(int i = 0; i < LTC_SIZE; ++i){
		vals_smooth[i] = round(cos(i*0.1) * 20);
		vals_rough[i] = round(cos(i) * 20);
	}

	start = When();
	for(int j = 0; j < 1000; ++j)
	{
		LTC<int, int, 3> comp;
		for(int i = 0; i < LTC_SIZE; ++i){
			bool a = comp.add(i, i%200);
			if(a)
				count_linear += 1;
		}
	}
	stop = When();
	cout << "Time (linear): " << (stop - start) << " (" << (((stop - start) / (LTC_SIZE * 1000)) * 1e9) << " ns/item)" << endl;
	start = When();
	for(int j = 0; j < 1000; ++j)
	{
		LTC<int, int, 3> comp;
		for(int i = 0; i < LTC_SIZE; ++i){
			bool a = comp.add(i, vals_smooth[i]);
			if(a)
				count_smooth += 1;
		}
	}
	stop = When();
	cout << "Time (smooth): " << (stop - start) << " (" << (((stop - start) / (LTC_SIZE * 1000)) * 1e9) << " ns/item)" << endl;

	start = When();
	for(int j = 0; j < 1000; ++j)
	{
		LTC<int, int, 3> comp;
		for(int i = 0; i < LTC_SIZE; ++i){
			bool a = comp.add(i, vals_rough[i]);
			if(a)
				count_rough += 1;
		}
	}
	stop = When();
	cout << "Time (rough): " << (stop - start) << " (" << (((stop - start) / (LTC_SIZE * 1000)) * 1e9) << " ns/item)" << endl;

	for(int i = 0; i < LTC_SIZE; ++i)
		vals_rand[i] = rand()%100000;

	start = When();
	for(int j = 0; j < 1000; ++j)
	{
		LTC<int, int, 3> comp;
		for(int i = 0; i < LTC_SIZE; ++i){
			bool a = comp.add(i, vals_rand[i]);
			if(a)
				count_rand += 1;
		}
	}
	stop = When();
	cout << "Time (random): " << (stop - start) << " (" << (((stop - start) / (LTC_SIZE * 1000)) * 1e9) << " ns/item)" << endl;
	cout << "Linear: " << ((count_linear)/1e6) << "M" << endl;
	cout << "Smooth: " << ((count_smooth)/1e6) << "M" << endl;
	cout << "Rough: " << ((count_rough)/1e6) << "M" << endl;
	cout << "Rand: " << ((count_rand)/1e6) << "M" << endl;
}
void test_reservoir_sampling(void){
	cout << "\t=== Reservoir Sampling ===" << endl;
	ReservoirSampling<int, 100, functions> rs;
	unsigned int sum_pre_count = 0;
	int pre_count[100] = {0}, count_loop = 1000000;
	pre_count[20] = 1000;
	pre_count[25] = 500;
	pre_count[50] = 400;
	pre_count[75] = 250;
	for(int i = 0; i < 100; ++i)
		sum_pre_count += pre_count[i];

	int count[100] = {0};
	double start, stop;
	start = When();
	for(int k = 0; k < count_loop; ++k)
		for(int i = 0; i < 100; ++i)
			for(int j = 0; j < pre_count[i]; ++j)
				rs.add(i);
	stop = When();
	cout << "Time: " << (stop - start) << " (" << (((stop - start) / (count_loop * sum_pre_count)) * 1e9 ) << " ns/item)" << endl;
	for(int i = 0; i < 100; ++i){
		int idx = rs[i];
		if(idx >= 0 && idx < 100)
			count[idx] += 1;
	}
	for(int i = 0; i < 100; ++i)
		if(count[i] < 0)
			cout << i << ": " << count[i] << endl;
}
#define MCNN_FEATURE_COUNT 4
//void test_mc_nn(void){
	//cout << "\t=== MCNN ===" << endl;
	//int dataset_size = 20;
	//MCNN<double, features_count, 10, 10> classifier;
	//double* dataset = (int*)calloc(features_count * dataset_size, sizeof(double));
	//int* labels = malloc(dataset_size * sizeof(int));
	//double* dataset2 = (int*)calloc(features_count * dataset_size, sizeof(double));
	//int* labels2 = malloc(dataset_size * sizeof(int));
	//for(int i = 0; i < dataset_size; ++i){
		//labels[i] = rand()%7;
		////TODO give some random value
		//for(int j = 0; j < features_count; ++j)
			//dataset[i][j] = 0;

		//labels2[i] = rand()%7;
		////TODO give some random value
		//for(int j = 0; j < features_count; ++j)
			//dataset2[i][j] = 0;
	//}
	
	//int const turn_count = 10;
	//for(int i = 0; i < turn_count; ++i){
		//for(int idx = 0; idx < dataset_size; ++idx)
			//classifier.train(dataset[idx], labels[idx]);	
		//for(int idx = 0; idx < dataset_size; ++idx)
			//classifier.train(dataset2[idx], labels2[idx]);	
	//}
	//classifier.print();
	//for(int idx = 8; idx < 13; ++idx){
		//auto predict = classifier.predict(dataset[idx]);
		//cout << "Prediction for " << idx << ": " << predict << endl;
		//classifier.train(dataset[idx], labels[idx]);	
		//classifier.print();
	//}
//}
int main(int argc, char** argv){
	test_ltc();
	test_reservoir_sampling();
	test_bloom();
	test_cuckoo();
	return 0;
}
